scipy/stats/_fit.py

```diff
@@ -711,3 +712,539 @@ def nlpsf(free_params, data=data):  # bind data NOW
+                                  'null_distribution'))
+
+
+def goodness_of_fit(dist, data, *, known_params=None, fit_params=None,
+                    guessed_params=None, statistic='ad', n_mc_samples=9999,
+                    random_state=None):
+    r"""
+    Perform a goodness of fit test comparing data to a distribution family.
+
+    Given a distribution family and data, perform a test of the null hypothesis
+    that the data were drawn from a distribution in that family. Any known
+    parameters of the distribution may be specified. Remaining parameters of
+    the distribution will be fit to the data, and the p-value of the test
+    is computed accordingly. Several statistics for comparing the distribution
+    to data are available.
+
+    Parameters
+    ----------
+    dist : `scipy.stats.rv_continuous`
+        The object representing the distribution family under the null
+        hypothesis.
+    data : 1D array_like
+        Finite, uncensored data to be tested.
+    known_params : dict, optional
+        A dictionary containing name-value pairs of known distribution
+        parameters. Monte Carlo samples are randomly drawn from the
+        null-hypothesized distribution with these values of the parameters.
+        Before the statistic is evaluated for each Monte Carlo sample, only
+        remaining unknown parameters of the null-hypothesized distribution
+        family are fit to the samples; the known parameters are held fixed.
+        If all parameters of the distribution family are known, then the step
+        of fitting the distribution family to each sample is omitted.
+    fit_params : dict, optional
+        A dictionary containing name-value pairs of distribution parameters
+        that have already been fit to the data, e.g. using `scipy.stats.fit`
+        or the ``fit`` method of `dist`. Monte Carlo samples are drawn from the
+        null-hypothesized distribution with these specified values of the
+        parameters, but these and all other unknown parameters of the
+        null-hypothesized distribution family are fit to the random samples
+        before the statistic is evaluated.
+    guessed_params : dict, optional
+        A dictionary containing name-value pairs of distribution parameters
+        which have been guessed. These parameters are always left free and
+        fit to the data - both the provided `data` and, later, samples drawn
+        from the null-hypothesized distribtion. However, the `guessed_params`
+        values will be provided as guesses to the fitting procedure.
+    statistic : {"ad", "ks", "cvm"}, optional
+        The statistic used to compare data to a distribution after fitting
+        unknown parameters of the distribution family to the data. The
+        Anderson-Darling ("ad"), Kolmogorov-Smirnov ("ks"), and
+        Cramer-von Mises ("cvm") statistics are available [1]_.
+    n_mc_samples : int, default: 9999
```

Comment:
Hopefully, you did not imagine that this choice was entirely arbitrary : ) It is the 9999 here because that is the default number of samples for all the [resampling methods](https://scipy.github.io/devdocs/reference/stats.html#resampling-methods) in SciPy.

One justification is that the denominator when calculating the p-value will be 9999 + 1 = 10000, so the p-value can always be represented with only four digits after the decimal point. This is preferable to a p-value with many digits after the decimal point because that would suggest that there is more precision than there really is. (That said, the p-values are still "exact" in the sense of [this paper](https://www.degruyter.com/document/doi/10.2202/1544-6115.1585/html).)

The primary reason is that [Art Owen](https://statistics.stanford.edu/people/art-b-owen) [suggested 9999 instead of 10,000](https://github.com/scipy/scipy/pull/13371#issuecomment-831341953) when I was writing `bootstrap`.

Reply:
💡